mentat/revisor/revisor.py

```diff
@@ -0,0 +1,132 @@
+    )
+    ctx.conversation.add_transcript_message(
+        ModelMessage(message=message, prior_messages=messages, message_type="revisor")
+    )
+
+    # Sometimes the model wraps response in a ```diff ``` block
+    # I believe new prompt fixes this but this makes sure it never interferes
+    if message.startswith("```diff\n"):
+        message = message[8:]
+        if message.endswith("\n```"):
+            message = message[:-4]
+
+    # This makes it more similar to a git diff so that we can use the pre existing git diff parser
+    message = "\n".join(message.split("\n")[2:])  # remove leading +++ and ---
+    post_diff = (
+        "diff --git a/file b/file\nindex 0000000..0000000\n--- a/file\n+++"
+        f" b/file\n{message}"
+    )
+    parsed_response = GitParser().parse_string(post_diff)
+
+    # Only modify the replacements of the current file edit
+    # (the new file edit doesn't know about file creation or renaming)
+    # Additionally, since we do this one at a time there should only ever be 1 file edit.
+    if parsed_response.file_edits:
+        stored_lines = _get_stored_lines(file_edit)
+        pre_lines = file_edit.get_updated_file_lines(stored_lines)
+        file_edit.replacements = parsed_response.file_edits[0].replacements
+        post_lines = file_edit.get_updated_file_lines(stored_lines)
+
+        diff_lines = difflib.unified_diff(pre_lines, post_lines, lineterm="")
+        diff_diff: List[str] = []
+        lexer = get_lexer(file_edit.file_path)
+        for line in diff_lines:
+            if line.startswith("---"):
+                diff_diff.append(f"{line}{file_edit.file_path}")
+            elif line.startswith("+++"):
+                new_name = (
+                    file_edit.rename_file_path
+                    if file_edit.rename_file_path is not None
+                    else file_edit.file_path
+                )
+                diff_diff.append(f"{line}{new_name}")
+            elif line.startswith("@@"):
+                diff_diff.append(line)
+            elif line.startswith("+"):
+                diff_diff.append(colored(line, "green"))
+            elif line.startswith("-"):
+                diff_diff.append(colored(line, "red"))
+            elif line.startswith(" "):
+                diff_diff.append(highlight_text(line, lexer))
+            else:
+                raise MentatError("Invalid Diff")
+        if diff_diff:
+            ctx.stream.send("Revision diff:", style="info")
+            ctx.stream.send(change_delimiter)
+            ctx.stream.send("\n".join(diff_diff))
+            ctx.stream.send(change_delimiter)
+        ctx.cost_tracker.display_last_api_call()
+
+
+async def revise_edits(file_edits: List[FileEdit]):
+    for file_edit in file_edits:
+        # We could do all edits asynchronously; of course, this risks getting rate limited and probably not worth effort
+        await revise_edit(file_edit)
```

Comment:
it might be worth the effort since it'd reduce latency a lot (and I've actually already implemented a balancer in a different repo) But yeah not for this PR, for now let's just try to get an experimental feature, not perfected yet